Add Red Hat OpenShift 2021-09-01-preview

mjudeikis opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [x] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:https://aka.ms/openapiportal
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit (including refactoring) and then push new changes, including version updates, in separate commits.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

Commit:
Adds base for updating Microsoft.RedHatOpenShift from version stable/2020-04-30 to version 2021-01-31-preview

Commit:
Updates readme

Commit:
Updates API version in new specs and examples

mjudeikis committed:
update ARO v2021-01-31-preview changes

mjudeikis committed:
Add azureresourceschema

mjudeikis committed:
systemData

mjudeikis committed:
update swagger

mjudeikis committed:
update readme

mjudeikis commented:
Same PR as in https://github.com/Azure/azure-rest-api-specs-pr/pull/4459 because I can't get anybody to look into it.

mjudeikis committed:
drop breaking change

mjudeikis commented:
Hey,

On the failures - I would like some guidance whats needs to be changed in the swagger itself to be compliant. Especially with sytemData failure as it looks like we have everything in there what is required. Would be nice if somebody would help to identify what is wrong, and not just saying its wrong.

weidongxu-microsoft commented:
I think the item is quite clear, you do not have `systemData` in `OpenShiftClusterUpdate`, why would it appear in your example json (OpenShiftClusters_Update.json)?

mjudeikis committed:
Add systemData

mjudeikis commented:
@weidongxu-microsoft added it again, and now getting error:
`READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST` 
Where its is same as here: https://github.com/Azure/azure-rest-api-specs/blob/88e7838a09868a51de3894114355c75929847a46/specification/mediaservices/resource-manager/Microsoft.Media/stable/2021-06-01/Accounts.json#L488-L492 

So what is different?

weidongxu-microsoft commented:
The `systemData` property is `readOnly`, which means client will not send it to service (there will be no `setter` method in client). And because client cannot set it, it should not be in the request. That is what the validation checks.

My idea is that your example JSON is wrong. Please double check whether service really need to read the `systemData` from client.

mjudeikis committed:
s/ListAdminKubeconfig/ListAdminCredentials

weidongxu-microsoft commented:
I don't know. I think mediaservices does not have `systemData` in their example JSON. https://github.com/Azure/azure-rest-api-specs/blob/master/specification/mediaservices/resource-manager/Microsoft.Media/stable/2021-06-01/examples/accounts-update.json

And in this example, `systemData` is not in request, but in response https://github.com/Azure/azure-rest-api-specs/blob/master/specification/mediaservices/resource-manager/Microsoft.Media/stable/2021-06-01/examples/jobs-update.json

mjudeikis commented:
> I don't know. I think mediaservices does not have `systemData` in their example JSON request. https://github.com/Azure/azure-rest-api-specs/blob/master/specification/mediaservices/resource-manager/Microsoft.Media/stable/2021-06-01/examples/accounts-update.json
> 
> And in this example, `systemData` is not in request, but in response https://github.com/Azure/azure-rest-api-specs/blob/master/specification/mediaservices/resource-manager/Microsoft.Media/stable/2021-06-01/examples/jobs-update.json

Got ya! I think I got this. Let me try to regenerate and see if this works. Thanks for helping here.. Really appreciate.

mjudeikis committed:
update systemData examples

mjudeikis committed:
add object type

mjudeikis commented:
@weidongxu-microsoft now looks like unrelated ci failures?

mjudeikis committed:
swap variable to preview

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,903 @@
+    "OpenShiftClusterAdminKubeconfig": {
+      "description": "OpenShiftClusterAdminKubeconfig represents an OpenShift cluster's admin kubeconfig.",
+      "properties": {
+        "kubeconfig": {
```
>kubeconfig [](http://example.com/codeflow?start=9&length=10)

If this is a secret, it should be tagged with x-ms-secret: true.

mjudeikis replied:
an you elaborate on how this looks like:
```
  "OpenShiftClusterAdminKubeconfig": {
      "description": "OpenShiftClusterAdminKubeconfig represents an OpenShift cluster's admin kubeconfig.",
      "type": "object",
      "properties": {
        "kubeconfig": {
          "description": "The base64-encoded kubeconfig file.",
          "type": "string",
          "x-ms-secret": true
        }
      }
    },
```

gives me on SDK generation test:
```
Loading AutoRest extension '@microsoft.azure/autorest.modeler' (2.3.38->2.3.38)
ERROR: Schema violation: Additional properties not allowed: x-ms-secret
    - file:///swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-09-01-preview/redhatopenshift.json:663:8 ($.definitions.OpenShiftClusterCredentials.properties.kubeadminPassword)
ERROR: Schema violation: Additional properties not allowed: x-ms-secret
    - file:///swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-09-01-preview/redhatopenshift.json:648:8 ($.definitions.OpenShiftClusterAdminKubeconfig.properties.kubeconfig)
FATAL: swagger-document/individual/schema-validator - FAILED
FATAL: Error: [OperationAbortedException] Error occurred. Exiting.
```
can it be our autorest image we use is old?

mentat9 replied:
Not sure: looks correct to me. @ArcturusZhang, any idea?

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,903 @@
+      ],
+      "type": "string",
+      "x-ms-enum": {
+        "name": "SDNProvider",
```
>SDNProvider [](http://example.com/codeflow?start=17&length=11)

Suggest expanding the SDN acroynm.

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,903 @@
+    "NetworkProfile": {
+      "description": "NetworkProfile represents a network profile.",
+      "properties": {
+        "sdnProvider": {
```
>sdnProvider [](http://example.com/codeflow?start=9&length=11)

I'm guessing SDN == Software Defined Network. ARM generally recommends against acronyms unless they are really common (e.g. IP, DNS). This is jargony enough I'd recommend spelling it out (softwareDefinedNetworkProvider). Not blocking ARM approval, just a suggestion.

mentat9 commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/redhatopenshift.json:
```diff
@@ -0,0 +1,903 @@
+    "SDNProvider": {
+      "description": "SDNProvider constants.",
+      "enum": [
+        "OVNKubernetes",
```
>OVNKubernetes [](http://example.com/codeflow?start=9&length=13)

Not sure whether it's correct, but if it is then [ Kubernetes, OpenShift ] seems better. If Kubernetes or OpenShift can have multiple different SDNs, then making these values specific is correct. Still probably best to expand the acronyms.

mjudeikis replied:
These are product names of its own and they are same as in the documentation. So would prefer to keep this.

msyyc commented:
For python, the CI failure is caused by wrong python configuration. I commit the right folder and it should work now.

mjudeikis committed:
Merge branch 'dev-redhatopenshift-Microsoft.RedHatOpenShift-2021-01-31-preview' of github.com:mjudeikis/azure-rest-api-specs into dev-redhatopenshift-Microsoft.RedHatOpenShift-2021-01-31-preview

mjudeikis committed:
Expand SDN & x-ms-secrets tag

mjudeikis commented:
@mentat9 Thansk or the review.
I expanded SDN name in the API. And I would prefer to keep (values as it is as they are the names)[https://docs.openshift.com/container-platform/4.8/networking/cluster-network-operator.html#nw-operator-cr-cno-object_cluster-network-operator]
```
Either OpenShiftSDN or OVNKubernetes. The cluster network provider is selected during installation. This value cannot be changed after cluster installation.
```

I added `type:object` to fields where it was complaining and not it is complaining that I added them :) 
In addition what should be right example handling for `x-ms-secrets` tags. We return these values in the body when POST is initiated. I was under impression this based on ARM spec how to handle secret.

weidongxu-microsoft commented:
Please check the Error in https://github.com/Azure/azure-rest-api-specs/pull/15616/checks?check_run_id=3310901502
I will ignore other Warnings about type, as it is OK to add `type: object`.

However the ModelValidation complains about secret, which I am not familiar. You would need to check with raychen or jianyxi

mjudeikis commented:
@raychen @jianyxi @mentat9  any ideas how `ms-x-secret` failures must be sorted? What is needed here?

mentat9 commented:
> @raychen @jianyxi @mentat9 any ideas how `ms-x-secret` failures must be sorted? What is needed here?

It looks like an issue in validation to me. ARM allows secrets in POST responses (only). The error is flagging your examples, which appear to only be referenced in the POST. SDK team, any idea?

mjudeikis commented:
@weidongxu-microsoft 
Error from your comment is what validation was asking to do, as previous failure were asking to add format, so we did.
```
   "diskSizeGB": {
          "description": "The disk size of the worker VMs.  Must be 128 or greater (immutable).",
          "type": "integer"
        },
```
to
```
        "diskSizeGB": {
          "format": "int64",
          "description": "The disk size of the worker VMs.",
          "type": "integer"
        },
```
@mentat9 Thanks, will do!

weidongxu-microsoft commented:
@jianyexi Does format change of adding `"format": "int64"` need approval from break changes board?

jianyexi commented:
> @jianyexi Jianye Xi FTE Does format change of adding `"format": "int64"` need approval from breaking changes board?

Yes

mjudeikis commented:
> > @jianyexi Jianye Xi FTE Does format change of adding `"format": "int64"` need approval from breaking changes board?
> 
> Yes

Does not adding it and keep old value and failing validation might be better option here? Please advise as this PR is becoming critical blocker :/

weidongxu-microsoft commented:
> > @raychen @jianyxi @mentat9 any ideas how `ms-x-secret` failures must be sorted? What is needed here?
> 
> It looks like an issue in validation to me. ARM allows secrets in POST responses (only). The error is flagging your examples, which appear to only be referenced in the POST. SDK team, any idea?

@jianyexi Please also take a look on this.

jianyexi commented:
> Does not adding it and keep old value and failing validation might be better option here? Please advise as this PR is becoming critical blocker :/
If you are fixing linting rule issue about the 'IntegerMustHaveType', It's recommend to fix it in new version 

for the validation of x-ms-secret , it's false positive, you can suppress it

mjudeikis commented:
> > Does not adding it and keep old value and failing validation might be better option here? Please advise as this PR is becoming critical blocker :/
> 
> If you are fixing linting rule issue about the 'IntegerMustHaveType', It's recommend to fix it in new version
> for the validation of x-ms-secret , it's false positive, you can suppress it

So what Im reading all our failures are false-positives and this should be good to merge? Or am I misreading?

weidongxu-microsoft commented:
@jianyexi
They are already on a new version, hence the Cross-Version Breaking Changes.
```
The new version has a different format than the previous one.
New: Microsoft.RedHatOpenShift/preview/2021-09-01-preview/redhatopenshift.json#L888:9
Old: Microsoft.RedHatOpenShift/stable/2020-04-30/redhatopenshift.json#L745:9
```

Also let me know whether there is schedule to fix the false alert on `x-ms-secret`.

@mjudeikis 
ARM reviewer has not yet approved this. Please review the comments.

jianyexi commented:
> @jianyexi Jianye Xi FTE
> They are already on a new version, hence the Cross-Version Breaking Changes.
@mjudeikis 
I see, then it's better to keep the old schema, and suppress the linting rule violation 'IntegerTypeMustHaveFormat'. 
And I approved the 'ModelValidation' CI failure, the CI check  is passed

mjudeikis commented:
> ARM reviewer has not yet approved this. Please review the comments.

I think Im missing what else needs to change?

mentat9 commented:
> > ARM reviewer has not yet approved this. Please review the comments.
> 
> I think Im missing what else needs to change?

Signed off from ARM side.

weidongxu-microsoft commented:
@mjudeikis 
OK, ARM just signed off. You can revert the change on integer format (which reported in Cross Version Breaking Changes), and we should be good.

mjudeikis committed:
revert int format

weidongxu-microsoft commented on specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-09-01-preview/redhatopenshift.json:
```diff
@@ -501,79 +554,122 @@
     },
+    "EncryptionAtHost": {
+      "description": "EncryptionAtHost represents encryption at host state",
+      "enum": [
+        "Disabled",
+        "Enabled"
+      ],
```
Please check if you need x-ms-enum
https://github.com/stankovski/AutoRest/blob/master/Documentation/swagger-extensions.md#x-ms-enum

mjudeikis replied:
We not added this to other "Two-state vars" Like, `Private`, `Public` or `Enable` `Disable`. I would like to keep this way, and if with the next version of API we will need, we can change  across the board.

mjudeikis replied:
If you say we need to I can add, but I would prefer to add it to all String type switches (more change)

weidongxu-microsoft replied:
`modelAsString` is actually your call, `false` or `true` is fine for SDK (though switch one to another going to be break changes).

The point is that you probably want to put a `name`. Else it could go unpredictable in SDK. This 2 `EncryptionAtHost` `EncryptionAtHost1` model in golang SDK might relate to the lack of `name`.
https://github.com/AzureSDKAutomation/azure-sdk-for-go/blob/078d46b49c7aaa98b99837e4755b4b9ecc40f51c/services/preview/mgmt/2021-09-01-preview/redhatopenshift/enums.go#L28-L56

mjudeikis replied:
Will trust in your judgment here as Im not super familiar with all this area :)
Added for both (Visibility and EncryptionAtHost)

weidongxu-microsoft replied:
Now it is only one of them https://github.com/AzureSDKAutomation/azure-sdk-for-go/blob/78a2b4b4486f787481f2fa9e51cd96356dbcea20/services/preview/mgmt/2021-09-01-preview/redhatopenshift/enums.go

I can merge when your service is ready.

weidongxu-microsoft reviewed (commented):
One comment about `x-ms-enum`.

Rest is good for me. Let me know when you need it merge (prefer to improve the `x-ms-enum`).

mjudeikis committed:
add x-ms-enum to Enryption and Visibility

mjudeikis commented:
@weidongxu-microsoft  good to merge
